# Challenge

In [1]:
import pandas as pd

from IPython.display import IFrame, display

## Question 1

In [2]:
formulation_path = "heater_model_documentation.pdf"
IFrame(formulation_path, width=700, height=1000)

## Question 2

In [3]:
%run -m simulate -i instances/example_bounds.toml -o results/example_bounds -s cbc -d -m lp
%run -m simulate -i instances/example_no_bounds.toml -o results/example_no_bounds -s cbc -d -m lp

Simulation of heating a building...
Welcome to the CBC MILP Solver 
Version: 2.10.8 
Build Date: May  7 2022 

command line - /opt/homebrew/bin/cbc -printingOptions all -import /var/folders/z1/q363rd3d7szg_187_9997w040000gn/T/tmpz0ohijle.pyomo.lp -stat=1 -solve -solu /var/folders/z1/q363rd3d7szg_187_9997w040000gn/T/tmpz0ohijle.pyomo.soln (default strategy 1)
Option for printingOptions changed from normal to all
Presolve 0 (-385) rows, 0 (-769) columns and 0 (-768) elements
Statistics for presolved model


Problem has 0 rows, 0 columns (0 with objective) and 0 elements
Column breakdown:
0 of type 0.0->inf, 0 of type 0.0->up, 0 of type lo->inf, 
0 of type lo->up, 0 of type free, 0 of type fixed, 
0 of type -inf->0.0, 0 of type -inf->up, 0 of type 0.0->1.0 
Row breakdown:
0 of type E 0.0, 0 of type E 1.0, 0 of type E -1.0, 
0 of type E other, 0 of type G 0.0, 0 of type G 1.0, 
0 of type G other, 0 of type L 0.0, 0 of type L 1.0, 
0 of type L other, 0 of type Range 0.0->1.0, 0 of type Rang

In [4]:
costs_bounds = pd.read_csv("results/example_bounds/objective_function.csv", index_col=0).squeeze()
costs_no_bounds = pd.read_csv("results/example_no_bounds/objective_function.csv", index_col=0).squeeze()
delta_costs = costs_no_bounds - costs_bounds
round(delta_costs, 2)

12.07

In [15]:
filepath_bounds = "results/example_bounds/cost.pdf"
IFrame(filepath_bounds, width=700, height=400)

In [16]:
filepath_no_bounds = "results/example_no_bounds/cost.pdf"
IFrame(filepath_no_bounds, width=700, height=400)

### Delta cost
The cost difference between adding bounds (21-26 degrees Celsius) and a fixed target temperature of 24 degrees Celsius is £12.07.

## Question 3

To account for non-deterministic ambient temperature and costs, one can resort to a stochastic optimisation formulation. Such a formulation can be cast as a fully deterministic linear program where different trajectories of temperatures and electricity costs (pre-coumputed) can be associated to the probability of each trajectory happening.

This formulation can potentially make the optimisation more computationally expensive as the number of decision variables is increased with each trajectory, increasing the complexity. On the other hand, since the optimisation takes into account several possible costs and temperatures, the results tend to be more robust (this approach is sometimes known as robust optimisation).